### PR TITLE
Prep

### PR DESCRIPTION
**Pull request by jshawl**

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -1 +1,115 @@
-# Functional Programming
\ No newline at end of file
+# Functional Programming
+
+## Learning Objectives
+
+- Identify the benefits of pure functions
+- Create a function that returns a function using closure.
```
Comment (nolds9) [inline readme.md]: why?

File: readme.md [hunk 2]
```diff
@@ -1 +1,115 @@
...
+- Explain what recursion is and why we use it
+- Explain the concept of a first-class function
```
Comment (nolds9) [inline readme.md]: is first-class function the same as higher-order function

File: readme.md [hunk 3]
```diff
@@ -1 +1,115 @@
...
+- Use `filter`, `map`, `reduce`, `sort`
+
```
Comment (nolds9) [inline readme.md:10]: - missing the concept of immutability

File: readme.md [hunk 4]
```diff
@@ -1 +1,115 @@
-# Functional Programming
\ No newline at end of file
+# Functional Programming
+
+## Learning Objectives
+
+- Identify the benefits of pure functions
```
Comment (nolds9) [inline readme.md]: no side effects -> why? how does that affect us?

**Review by nolds9 — COMMENTED** [review]
First pass

File: readme.md (hunk 1 above)
Comment (jshawl) [reply readme.md]: removed it

File: readme.md [hunk 5]
```diff
@@ -1 +1,258 @@
...
+# Functional Programming
+
+## Learning Objectives
+
+- Identify the benefits of pure functions and avoiding side effects
+- Define immutability and how it relates to pure functions
+- Explain what recursion is and why we use it
+- Explain the concepts of first class and higher order functions
+- Use `filter`, `map`, `reduce`, `sort`
+
```
Comment (nolds9) [inline readme.md:10]: ❤️ these LOs

File: readme.md [hunk 6]
```diff
@@ -1 +1,258 @@
...
+## Framing
+
+### Why functional programming? Why now?
+
```
Comment (nolds9) [inline readme.md:14]: Let's talk a little bit about paradigms in programming....

Could turn it into a quick STWG - "What are some examples of paradigms we've followed so far in this class?"

File: readme.md [hunk 7]
```diff
@@ -1 +1,258 @@
...
+  - privatizing the complexity
+- inherit
+  - sharing the complexity
+
+Object oriented programming reminds me of an LCD display. If one pixel goes out,
+no problem. It doesn't affect the rest of the system. It's easy to identify where
+something's broken.
+
+>The problem with object-oriented languages is they’ve got all this implicit environment that they carry around with them. You wanted a banana but what you got was a gorilla holding the banana and the entire jungle. — Joe Armstrong
+
```
Comment (nolds9) [inline readme.md:35]: for example...

File: readme.md [hunk 8]
```diff
@@ -1 +1,258 @@
...
+## Pure Functions (stateless)
+
```
Comment (nolds9) [inline readme.md:52]: Let's look at another high level concept in programming...

File: readme.md [hunk 9]
```diff
@@ -1 +1,258 @@
...
+Here's an example of an impure function:
+
+```js
+var age = 27
+function increaseAgeBy(int){
+  return age += int
+}
+increaseAgeBy(2)
+// how can we demonstrate the impurity here?
+```
```
Comment (nolds9) [inline readme.md:67]: STWG - what makes this function impure?

File: readme.md [hunk 10]
```diff
@@ -1 +1,258 @@
...
+
+```js
+var age = 27
+function increaseAgeBy(age,int){
+  return age += int
+}
+increaseAgeBy(2)
+// how can we demonstrate the purity here?
+```
+
```
Comment (nolds9) [inline readme.md:90]: could tie in these examples to side effects

File: readme.md [hunk 11]
```diff
@@ -1 +1,258 @@
...
+Modify `reverse` so that the original `nums` remains in the correct order.
+
+### Immutability
+
+In the above example, `collection` is a __mutable__ data structure. `Mutable`
+objects are objects whose state can be modified after creation.
+
+When working with pure functions, it is important to avoid changing (or mutating)
+objects outside of the function.
+
```
Comment (nolds9) [inline readme.md:116]: maybe a quick example, but should be clear from the previous example....

File: readme.md [hunk 12]
```diff
@@ -1 +1,258 @@
...
+## Higher Order && First Class Functions
+
```
Comment (nolds9) [inline readme.md:118]: If you come from a programming background besides JS, you might be familiar with the concepts of Higher Order and First Class Functions....if not, let's look at what these terms mean.

File: readme.md [hunk 13]
```diff
@@ -1 +1,258 @@
...
+1. find
+1. every
+1. some
+
+Once in your groups...
+
+1. Create an explain it to me like i'm 5 explanation
+1. Create a codpen / jsfiddle demoing how it works
+1. Come up with a way to remember what it is and what it does
+
```
Comment (nolds9) [inline readme.md:173]: maybe includes links for where to find out more

File: readme.md [hunk 14]
```diff
@@ -1 +1,258 @@
...
+```js
+function fib(n){
+  if(n === 1 || n === 2){
+    return 1
+  }
+  return fib(n-1) + fib(n-2)
+}
+```
+
```
Comment (nolds9) [inline readme.md:213]: why use recursion?

Comment (nolds9) [reply readme.md:213]: Some problems can only be solved with recursion...

File: readme.md [hunk 15]
```diff
@@ -1 +1,258 @@
...
+Here's a sample nav menu:
+
+```js
+document.body.appendChild(ul(
+  li(a('Home','/')),
+  li(a('About','/about')),
+  li(a('Contact Us','/contact'))
+))
+
+function ul(...children){
```
Comment (nolds9) [inline readme.md:264]: don't think I'm going to use the spread operator in the es6 lesson fyi...

**Review by nolds9 — COMMENTED** [review]
LGTM

**Pull request merged**